### PR TITLE
ceph-dedup-tool: add new FastCDC chunker, and make estimate test a range of chunk sizes

### PR DESCRIPTION
**Pull request by liewegas**
- abstract CDC (content-defined chunking) interface
- new FastCDC chunker that's faster than rabin (also, our rabin implementation doesn't quite work)
- dedup tool estimate will test a range by chunk sizes by default so you can run it with minimal arguments and get back information about how dedup-able a pool is
- some dedup tool cleanup

### CODE REVIEW

**Comment by athanatos** [comment]
Is the idea to reenable rabin chunking once it passes the relevant test, or do we expect FastCDC to replace it?

I'll take a look at the paper this evening and have a look at the FastCDC files, the rest looks good though.

**Comment by mattbenjamin** [comment]
@mykaul fyi

**Comment by liewegas** [comment]
> Is the idea to reenable rabin chunking once it passes the relevant test, or do we expect FastCDC to replace it?
> 
> I'll take a look at the paper this evening and have a look at the FastCDC files, the rest looks good though.

I'm not sure there is much point in fixing rabin... AFAICS FastCDC is superior.  I did a quick literature search and found a few other interesting papers but nothing that is a drop-in replacement:

https://dl.acm.org/doi/pdf/10.1145/3357223.3362731
https://dl.acm.org/doi/pdf/10.1145/3319647.3325834

(from https://scholar.google.com/scholar?cites=11139832257993135599&as_sdt=5,50&sciodt=0,50&hl=en)

File: src/common/FastCDC.cc [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	fp = (fp << 1) ^ table[*(unsigned char*)pp];
+      }
+    }
+    ceph_assert(pos < len);
+
+    // find an end marker
+    if (_scan(&p, &pp, &pe, pos,  // for the first "small" region
+	      std::min(len,
+		       cstart + (1 << (target_bits - TARGET_WINDOW_BITS))),
+	      fp, small_mask, table) &&
```
Comment (myoungwon) [inline src/common/FastCDC.cc:146]: Just question. What is the purpose of this?
In the paper, 
>for ; i < NormalSize; i++; do
fp = (fp << 1) + Gear[ src[i] ];
if ! ( fp & MaskS ) then
return i; //if the masked bits are all ‘0’
for ; i<n; i++; do
fp = (fp << 1) + Gear[ src[i] ];
if ! ( fp & MaskL ) then
return i;

It seems that below two conditions ((target_bits + TARGET_WINDOW_BITS) and (1 << max_bits))
match both MaskS and MaskL..

Comment (athanatos) [reply src/common/FastCDC.cc:146]: I'm not quite sure what you mean.

**Comment by athanatos** [comment]
Additional changes look good.

File: src/common/FastCDC.cc (hunk 1 above)
Comment (liewegas) [reply src/common/FastCDC.cc:146]: I changed things a bit from the paper.  For the original FastCDC, anything < target_size uses maskS, and anything > uses maskL.  This implementation uses maskS < target_size/2, mask for target_size/2 to target_size*2, and maskL above that.

Comment (myoungwon) [reply src/common/FastCDC.cc:146]: Ok. It means the normalization level is 3?

Comment (liewegas) [reply src/common/FastCDC.cc:146]: Normalization level is currently 2:
```
//  how many bits to set/clear in the small/large masks
#define TARGET_WINDOW_MASK_BITS  2
```

Comment (liewegas) [reply src/common/FastCDC.cc:146]: I will update the source to describe how it corresponds to the paper's terms.

Comment (liewegas) [reply src/common/FastCDC.cc]: I did a quick test (not super comprehensive, but something) comparing target_windwo_bits 1 vs 0 (==original fastcdc) and 1 performed slightly better.  Probably need to use a more representative data set to be sure, though.

**Comment by liewegas** [comment]
@myoungwon okay if i just remove rabin?  I'm not sure it's worth debugging to sort out why it's not working properly.

**Comment by myoungwon** [comment]
OK. Rabin has many options to consider. So, it is difficult.

**Comment by tchaikov** [comment]
@liewegas needs rebase

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-05-23_03:49:52-rados-wip-kefu-testing-2020-05-23-0054-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-05-27_02:18:21-rados-wip-kefu-testing-2020-05-23-0054-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-05-27_06:49:52-rados-wip-kefu-testing-2020-05-27-1240-distro-basic-smithi/

the test failures are either tracked or fixed by

- https://tracker.ceph.com/issues/45519
- #33272

**Pull request merged**